sylt/sylt-common/src/flat_value.rs

```diff
@@ -0,0 +1,180 @@
+            ),
+            Value::Float(f) => FlatValue::Float(*f),
+            Value::Int(i) => FlatValue::Int(*i),
+            Value::Bool(b) => FlatValue::Bool(*b),
+            Value::String(s) => FlatValue::String(String::clone(s)),
+            Value::Function(captured, ty, slot) => FlatValue::Function(
+                captured
+                    .iter()
+                    .map(|upvalue| FlatUpValue { slot: upvalue.borrow().slot, value: Self::pack_inner(&upvalue.borrow().value, pack, seen) })
+                    .collect(),
+                ty.clone(),
+                *slot,
+            ),
+            Value::ExternFunction(slot) => FlatValue::ExternFunction(*slot),
+            Value::Unknown => FlatValue::Unknown,
+            Value::Nil => FlatValue::Nil,
+            Value::Union(_) => {
+                unreachable!("Cannot send union values over the network");
+            }
+        };
+        pack[id] = val;
+        id
+    }
+
+    fn partial_unpack(value: FlatValue) -> Value {
+        match value {
+            FlatValue::Field(s) => Value::Field(s),
+            FlatValue::Ty(ty) => Value::Ty(ty),
+            FlatValue::Blob(slot) => Value::Blob(slot),
+            FlatValue::Instance(ty_slot, _) => Value::Instance(
+                ty_slot,
+                Rc::new(RefCell::new(HashMap::new())),
+            ),
+            FlatValue::Tuple(_) => Value::Tuple(Rc::new(Vec::new())),
+            FlatValue::List(_) => Value::List(Rc::new(RefCell::new(Vec::new()))),
+            FlatValue::Set(_) => Value::Set(Rc::new(RefCell::new(HashSet::new()))),
+            FlatValue::Dict(_) => Value::Dict(Rc::new(RefCell::new(HashMap::new()))),
+            FlatValue::Float(f) => Value::Float(f),
+            FlatValue::Int(i) => Value::Int(i),
+            FlatValue::Bool(b) => Value::Bool(b),
+            FlatValue::String(s) => Value::String(Rc::new(s)),
+            FlatValue::Function(_, ty, slot) => Value::Function(
+                Rc::new(Vec::new()),
+                ty,
+                slot,
+            ),
+            FlatValue::ExternFunction(slot) => Value::ExternFunction(slot),
+            FlatValue::Unknown => Value::Unknown,
+            FlatValue::Nil => Value::Nil,
+        }
+    }
+
+    pub fn unpack(pack: &Vec<FlatValue>) -> Value {
+        let mut mapping: Vec<Value> = pack.iter().cloned().map(Self::partial_unpack).collect();
+        for (i, x) in mapping.iter().enumerate().rev() {
+            match (&pack[i], x) {
+                (FlatValue::Instance(_, flat), Value::Instance(_, values)) => {
+                    let mut values = values.borrow_mut();
+                    for (key, id) in flat {
+                        values.insert(key.clone(), mapping[*id].clone());
+                    }
+                }
+                (FlatValue::Tuple(flat), Value::Tuple(values)) => {
+                    let values = unsafe { (Rc::as_ptr(values) as *mut Vec<Value>).as_mut() }.unwrap();
```

Comment:
This is potentially VERY WRONG and SHOULD NOT BE TRUSTED. It will be rewritten in an actual implementation later.

Reply:
It feels right though.